### PR TITLE
feat(stepper): webcomponent

### PR DESCRIPTION
**Pull request by mJarsater**
**Describe pull-request**  
Web component version of the stepper component and its children.

**Solving issue**  
Fixes: [AB#2958](https://dev.azure.com/scaniadigitaldesignsystem/3d9eb9c6-0045-473d-bc1d-e842dd779200/_workitems/edit/2958)

**How to test**  
1. Go to storybook link below
2. Check in Components -> Stepper -> Web component
3. Try out the controls and see that the design matches the one in figma.

**Suggested test steps**
- [ ] Browser testing (Chrome, Safari, Firefox) 
- [ ] Keyboard operability
- [ ] Interactive elements have labels.
- [ ] Storybook controls
- [ ] Design/controls/props is aligned with other components 
- [ ] Dark/light mode and variants 
- [ ] Input fields – values should be displayed properly 
- [ ] Events

**Screenshots**  
_If applicable, add screenshots to help explain_

**Additional context**  
_Add any other context about the pull-request here._

### CODE REVIEW

**Review by theJohnnyMe — DISMISSED** [review]
Some small things:

1. Icon in Figma for Error state seems to be different: 
Screenshot 2023-01-04 at 11 12 17

2. Please link both "stepper" and "stepper item" in the "Notes tab
3. Maybe leave only "Native" and "WebComponent" stories?

File: tegel/src/components/stepper/stepper-component.tsx [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  tag: 'sdds-stepper',
+  styleUrl: 'stepper-component.scss',
+  shadow: true,
+})
+export class SddsStepper {
+  /** The direction the children are layed out. */
+  @Prop() direction: 'horizontal' | 'vertical' = 'horizontal';
+
+  /** Text position, only available on direction:horizontal */
+  @Prop() textPosition: 'aside' | 'below' = 'below';
```
Comment (EmelieLitwin) [inline tegel/src/components/stepper/stepper-component.tsx:13]: Maybe call the prop labelPosition to be more like other components + it's called labels in Figma.
Would be nice to have `@Prop() labelPosition: 'aside' | 'below' | 'none' = 'below';` or something to be more consistent with other components but maybe that's hard if it doesn't work that way for vertical

Comment (mJarsater) [reply tegel/src/components/stepper/stepper-component.tsx:13]: Totally agree with the prop name! I guess the only thing with having 'none' as one of the potential attributes is that it's not really label position. I'll change the other stuff though, and we can discuss the attribute more 😄

**Comment by EmelieLitwin** [comment]
@charlottelorensson You are doing stepper documentation soon, right? Maybe coordinate the storybook example with that? My thinking was that the inactive step should be the last to show it's upcoming.
I gave the feedback that "active" should be "current" because I think that is clearer if that is what it's supposed to be. And in that case, it should be renamed in Figma as well. Maybe inactive should be something else as well, I don't know. Is error used as incomplete and inactive only for upcoming?

**Comment by mJarsater** [comment]
> One thing I remember was discussed previously regarding the stepper was that the lines in between the steps should be the same length even when some step names are longer. It's not updated in Figma yet but would that be possible to add here? Or should it be made in Figma first?

No problem! How long should the line be?

**Comment by charlottelorensson** [comment]
> > One thing I remember was discussed previously regarding the stepper was that the lines in between the steps should be the same length even when some step names are longer. It's not updated in Figma yet but would that be possible to add here? Or should it be made in Figma first?
> 
> No problem! How long should the line be?
> 
> And how should the text be displayed if it is very long, like in this case: Screenshot 2023-02-03 at 12 27 25

If possible it would use the same length on the line to the longest word on all of them. However, since there's no documentation on it yet I'm not sure where the upper limit goes. Now it just continues. Ideally the naming would be done in such a way so that they are not super long. Maybe it would sometimes need to use 2 lines. This would have to be explored a bit further once picked up.

**Comment by mJarsater** [comment]
> > > One thing I remember was discussed previously regarding the stepper was that the lines in between the steps should be the same length even when some step names are longer. It's not updated in Figma yet but would that be possible to add here? Or should it be made in Figma first?
> > 
> > 
> > No problem! How long should the line be?
> > And how should the text be displayed if it is very long, like in this case: Screenshot 2023-02-03 at 12 27 25
> 
> If possible it would use the same length on the line to the longest word on all of them. However, since there's no documentation on it yet I'm not sure where the upper limit goes. Now it just continues. Ideally the naming would be done in such a way so that they are not super long. Maybe it would sometimes need to use 2 lines. This would have to be explored a bit further once picked up.

Alright, no problem. How should we continue with this PR for now? Should we merge it as it is, or do we want to hold off and wait for the design?

**Comment by charlottelorensson** [comment]
> > > > One thing I remember was discussed previously regarding the stepper was that the lines in between the steps should be the same length even when some step names are longer. It's not updated in Figma yet but would that be possible to add here? Or should it be made in Figma first?
> > > 
> > > 
> > > No problem! How long should the line be?
> > > And how should the text be displayed if it is very long, like in this case: Screenshot 2023-02-03 at 12 27 25
> > 
> > 
> > If possible it would use the same length on the line to the longest word on all of them. However, since there's no documentation on it yet I'm not sure where the upper limit goes. Now it just continues. Ideally the naming would be done in such a way so that they are not super long. Maybe it would sometimes need to use 2 lines. This would have to be explored a bit further once picked up.
> 
> Alright, no problem. How should we continue with this PR for now? Should we merge it as it is, or do we want to hold off and wait for the design?

Not sure what would be the best thing to do. A version could be to see if we could make the line length behave as suggested and then leave the word length to be handled in the documentation for now? What do you think?

**Comment by mJarsater** [comment]
> > > > > One thing I remember was discussed previously regarding the stepper was that the lines in between the steps should be the same length even when some step names are longer. It's not updated in Figma yet but would that be possible to add here? Or should it be made in Figma first?
> > > > 
> > > > 
> > > > No problem! How long should the line be?
> > > > And how should the text be displayed if it is very long, like in this case: Screenshot 2023-02-03 at 12 27 25
> > > 
> > > 
> > > If possible it would use the same length on the line to the longest word on all of them. However, since there's no documentation on it yet I'm not sure where the upper limit goes. Now it just continues. Ideally the naming would be done in such a way so that they are not super long. Maybe it would sometimes need to use 2 lines. This would have to be explored a bit further once picked up.
> > 
> > 
> > Alright, no problem. How should we continue with this PR for now? Should we merge it as it is, or do we want to hold off and wait for the design?
> 
> Not sure what would be the best thing to do. A version could be to see if we could make the line length behave as suggested and then leave the word length to be handled in the documentation for now? What do you think?

Sure thing! You can have a look at the preview now if you want to. Should be updated with all the latest changes.

https://deploy-preview-762--tegel-storybook.netlify.app/?path=/story/components-stepper--web-component

**Comment by charlottelorensson** [comment]
> > > > > > One thing I remember was discussed previously regarding the stepper was that the lines in between the steps should be the same length even when some step names are longer. It's not updated in Figma yet but would that be possible to add here? Or should it be made in Figma first?
> > > > > 
> > > > > 
> > > > > No problem! How long should the line be?
> > > > > And how should the text be displayed if it is very long, like in this case: Screenshot 2023-02-03 at 12 27 25
> > > > 
> > > > 
> > > > If possible it would use the same length on the line to the longest word on all of them. However, since there's no documentation on it yet I'm not sure where the upper limit goes. Now it just continues. Ideally the naming would be done in such a way so that they are not super long. Maybe it would sometimes need to use 2 lines. This would have to be explored a bit further once picked up.
> > > 
> > > 
> > > Alright, no problem. How should we continue with this PR for now? Should we merge it as it is, or do we want to hold off and wait for the design?
> > 
> > 
> > Not sure what would be the best thing to do. A version could be to see if we could make the line length behave as suggested and then leave the word length to be handled in the documentation for now? What do you think?
> 
> Sure thing! You can have a look at the preview now if you want to. Should be updated with all the latest changes.
> 
> https://deploy-preview-762--tegel-storybook.netlify.app/?path=/story/components-stepper--web-component

Awesome, nice job, thank you!

File: tegel/src/components/stepper/stepper-item/readme.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+| ----------- | ------------ | -------------------------------- | ------------------------------------------------- | ------------ |
+| `labelText` | `label-text` | Label text for the stepper-item. | `string`                                          | `''`         |
+| `state`     | `state`      | State of the stepper-item        | `"current" \| "error" \| "success" \| "upcoming"` | `'upcoming'` |
+
+
+## Methods
+
+### `setWidth(width: any) => Promise<void>`
+
+Method to set the width if the stepper item based on its siblings widht, used by the parent element.
```
Comment (timrombergjakobsson) [inline tegel/src/components/stepper/stepper-item/readme.md:20]: Super small thing, but incorrect spelling on "width" here(I do it also all the time.)

Comment (mJarsater) [reply tegel/src/components/stepper/stepper-item/readme.md:20]: Nice catch 🎣

File: tegel/src/components/stepper/stepper-item/stepper-item.tsx [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+  componentWillLoad() {
+    this.stepperEl = this.el.closest('sdds-stepper');
+    this.direction = this.stepperEl.direction;
+    this.labelPosition = this.stepperEl.labelPosition;
+    this.size = this.stepperEl.size;
+    this.hideLabel = this.stepperEl.hideLabels;
+  }
+
+  /** Method to set the width if the stepper item based on its siblings widht, used by the parent element. */
```
Comment (mJarsater) [inline tegel/src/components/stepper/stepper-item/stepper-item.tsx:37]:
```suggestion
  /** Method to set the width if the stepper item based on its siblings width, used by the parent element. */
```

**Review by nathalielindqvist — COMMENTED** [review]
Some mighty fine work! 🔥 Just some minor things that could use an extra look, great job! 😄

File: tegel/src/components/stepper/stepper-webcomponent.stories.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+      control: {
+        type: 'radio',
+      },
+      table: {
+        type: { summary: 'string' },
+        defaultValue: { summary: 'lg' },
+      },
+      options: ['Large', 'Small'],
+    },
+    hideLabels: {
```
[the remark this answers is not in the document]
Comment (mJarsater) [reply tegel/src/components/stepper/stepper-webcomponent.stories.ts:61]: Will do!

File: tegel/src/components/stepper/stepper-item/stepper-item.scss [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+        font: var(--sdds-detail-01);
+        letter-spacing: var(--sdds-detail-01-ls);
+      }
+
+      &.sm {
+        font: var(--sdds-detail-05);
+        letter-spacing: var(--sdds-detail-05-ls);
+      }
+
+      &.upcoming {
```
[the remark this answers is not in the document]
Comment (mJarsater) [reply tegel/src/components/stepper/stepper-item/stepper-item.scss:206]: Nice, had totally forgotten about dark mode for this one! 🕶️

File: tegel/src/components/stepper/sdds-stepper.stories.ts [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+  },
+  argTypes: {
+    direction: {
+      name: 'Direction',
+      description: 'The direction which the stepper is displayed.',
+      control: {
+        type: 'radio',
+      },
+      table: {
+        type: { summary: 'string' },
```
Comment (nathalielindqvist) [inline tegel/src/components/stepper/sdds-stepper.stories.ts:31]: Do we need to have the 'type' in the table summary when we have the type in control? Looks a bit off in the description :)
image

File: tegel/src/components/stepper/sdds-stepper.stories.ts [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        defaultValue: { summary: 'lg' },
+      },
+      options: ['Large', 'Small'],
+    },
+    hideLabels: {
```
Comment (nathalielindqvist) [inline tegel/src/components/stepper/sdds-stepper.stories.ts:61]: Is there a reason why you didn't add a description? ☺️

File: tegel/src/components/stepper/sdds-stepper.stories.ts (hunk 6 above)
Comment (mJarsater) [reply tegel/src/components/stepper/sdds-stepper.stories.ts:31]: Ideally I would have it in both places and specify enum types also (where we have those), but it's something we can look into in the future. So I'll remove it for now 😄

File: tegel/src/components/stepper/sdds-stepper.stories.ts (hunk 7 above)
Comment (mJarsater) [reply tegel/src/components/stepper/sdds-stepper.stories.ts:58]: Added it now 😄

**Review by nathalielindqvist — APPROVED** [review]
LGTM! 🔥

**Review by theJohnnyMe — APPROVED** [review]
Nice work Martin! 🤘

**Pull request merged**